<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Matthew Stephens" />

<meta name="date" content="2020-07-22" />

<title>Generate data from a simple genetic mixture model</title>

<script src="site_libs/header-attrs-2.9/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">fiveMinuteStats</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephens999/fiveMinuteStats">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Generate data from a simple genetic mixture model</h1>
<h4 class="author">Matthew Stephens</h4>
<h4 class="date">2020-07-22</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-01-16
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>fiveMinuteStats/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed12345code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(12345)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed12345code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(12345)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomKiseokUchicagofiveMinuteStatstree4aba20d757e4fa363d1ba34e5de1f79b6b1d530ftargetblank4aba20da"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/KiseokUchicago/fiveMinuteStats/tree/4aba20d757e4fa363d1ba34e5de1f79b6b1d530f" target="_blank">4aba20d</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomKiseokUchicagofiveMinuteStatstree4aba20d757e4fa363d1ba34e5de1f79b6b1d530ftargetblank4aba20da" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/KiseokUchicago/fiveMinuteStats/tree/4aba20d757e4fa363d1ba34e5de1f79b6b1d530f" target="_blank">4aba20d</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/

Unstaged changes:
    Modified:   analysis/HW1_Basic_Monte_Carlo_and_Bayesian_inference.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/220113_r_simplemix.Rmd</code>) and HTML (<code>docs/220113_r_simplemix.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/KiseokUchicago/fiveMinuteStats/blob/4aba20d757e4fa363d1ba34e5de1f79b6b1d530f/analysis/220113_r_simplemix.Rmd" target="_blank">4aba20d</a>
</td>
<td>
Kiseok Lee
</td>
<td>
2022-01-16
</td>
<td>
wflow_publish(“analysis/220113_r_simplemix.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This vignette is an “ice-breaker”, to motivate learning and statistical inference centered around genetic mixtures.</p>
<p>You don’t need to know what a mixture model is to understand this, but if you want to know more about mixture models in general you might read this <a href="intro_to_mixture_models.html">introduction to mixture models</a>.</p>
</div>
<div id="simulating-haploid-data-for-a-single-population" class="section level2">
<h2>Simulating haploid data for a single population</h2>
<p>First we consider simulating genotype data for <span class="math inline">\(n\)</span> haploid individuals at <span class="math inline">\(R\)</span> independent bi-allelic loci (positions along the genome) sampled from a population.</p>
<p>The term “haploid” means that each individual has only one copy of their genome. (Most animals, are diploid, which means they have two copies of their genome – one inherited from the mother and the other inherited from the father. However, focussing on haploid individuals makes the ideas and code easier to follow. Once you understand the haploid case it is not too hard to extend the ideas to the diploid case.)</p>
<p>The term “bi-allelic” means that the loci have two possible alleles (types), which for convenience we will label <code>0</code> and <code>1</code>.</p>
<p>Under these assumptions, the genotype for each individual is simply a sequence of 0s and 1s. The probability of seeing a <code>0</code> vs a <code>1</code> at each locus is determined by the “allele frequencies” at each locus, which we will specify by a vector <code>p</code>. Specifically, <code>p[r]</code> specifies the frequency of the <code>1</code> allele at locus <code>r</code>.</p>
<p>The following code simulates from this model.</p>
<pre class="r"><code>#&#39; @param n number of samples
#&#39; @param p a vector of allele frequencies
r_haploid_genotypes = function(n,p){
  R = length(p)
  x = matrix(nrow = n, ncol=R)
  for(i in 1:n){
    x[i,] = rbinom(R,rep(1,R),p)
  }
  return(x)
}</code></pre>
<p>KS - trying to figure out what rbinom function is doing</p>
<pre class="r"><code>p = seq(0.1,0.9,length=9)
R = length(p)
R</code></pre>
<pre><code>[1] 9</code></pre>
<pre class="r"><code>rep(1,R)</code></pre>
<pre><code>[1] 1 1 1 1 1 1 1 1 1</code></pre>
<pre class="r"><code>rbinom(9, 1, 0.9)</code></pre>
<pre><code>[1] 1 1 1 1 1 1 1 1 1</code></pre>
<pre class="r"><code>rbinom(9, rep(1,10), c(.1,.1,.1,.1,.1,.1,.1,.1,.1,.1,.1))</code></pre>
<pre><code>[1] 1 0 0 0 0 0 0 0 0</code></pre>
<pre class="r"><code>rbinom(R,rep(1,R),p)</code></pre>
<pre><code>[1] 0 1 0 0 1 0 1 1 1</code></pre>
<div id="example" class="section level3">
<h3>Example</h3>
<p>To illustrate this function we simulate a small example dataset, containing 20 individuals at 9 loci. The frequencies of the <code>1</code> allele at the loci are increasing from 0.1 at the first locus to 0.9 at the 9th locus (the pattern is not supposed to be realistic, it is just to help illustrate the idea).</p>
<p>As you can see, the <code>1</code> allele is rarer at the earlier loci, whereas the <code>0</code> allele is rarer at the later loci.</p>
<pre class="r"><code>set.seed(123)
p = seq(0.1,0.9,length=9)
x = r_haploid_genotypes(20,p)
p</code></pre>
<pre><code>[1] 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9</code></pre>
<pre class="r"><code>x </code></pre>
<pre><code>      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9]
 [1,]    0    0    0    1    1    1    1    0    1
 [2,]    0    1    0    1    1    1    0    1    1
 [3,]    0    1    1    1    1    0    1    1    1
 [4,]    0    0    0    1    1    0    0    1    1
 [5,]    0    0    0    0    0    1    1    1    1
 [6,]    0    0    0    0    1    1    1    1    1
 [7,]    0    0    0    1    1    1    1    1    1
 [8,]    0    1    0    1    1    0    1    1    1
 [9,]    0    0    0    0    0    0    1    1    1
[10,]    0    0    1    0    0    0    0    0    1
[11,]    0    0    0    1    0    1    0    1    1
[12,]    0    0    0    0    1    1    0    0    1
[13,]    0    0    1    0    0    0    0    1    1
[14,]    1    0    0    1    0    1    1    1    0
[15,]    0    0    0    1    1    0    1    1    1
[16,]    0    1    1    1    0    1    1    1    1
[17,]    0    0    0    0    0    0    0    1    1
[18,]    0    0    0    0    0    1    1    1    1
[19,]    0    0    0    0    1    1    1    1    1
[20,]    0    1    1    1    1    1    1    0    1</code></pre>
</div>
</div>
<div id="simulating-haploid-data-from-a-mixture-of-two-populations" class="section level2">
<h2>Simulating haploid data from a mixture of two populations</h2>
<p>Now suppose we sample from a group of individuals formed by mixing together the individuals from two different populations. This is an example of a “mixture model”.</p>
<p>For simplicity we will assume the two different populations are mixed in equal proportions. That is the “mixture proportions” are 0.5 and 0.5.</p>
<p>The following <code>r_simplemix</code> function generates data from such a model. The allele frequencies for the two populations must be specified in a matrix <code>P</code> whose first row contains the allele frequencies for population <code>1</code> and second row is the allele frequencies for population <code>2</code>. (So element <code>P[k,r]</code> is the frequency of the <code>1</code> alleles in population <code>k</code> at locus <code>r</code>.) For each individual <code>i</code> it randomly samples the population (<code>z[i]</code>) to be 1 or 2, and then uses the <code>r_haploid_genotypes</code> to generate the genotypes (<code>x[i,]</code>) from that population.</p>
<pre class="r"><code>#&#39; @param n number of samples
#&#39; @param P a 2 by R matrix of allele frequencies
r_simplemix = function(n,P){
  R = ncol(P) # number of loci
  z = rep(0,n) # used to store population of origin of each individual
  x = matrix(nrow = n, ncol=R) #used to store genotypes
  
  for(i in 1:n){
    z[i] = sample(1:2, size=1, prob=c(0.5,0.5))
    x[i,] = r_haploid_genotypes(1,P[z[i],])
  }
  return(list(x=x,z=z))
}</code></pre>
<p>Modification for exercise 1</p>
<pre class="r"><code>#&#39; @param n number of samples
#&#39; @param P a 2 by R matrix of allele frequencies
r_simplemix_w = function(n, P, w){
  R = ncol(P) # number of loci
  z = rep(0,n) # used to store population of origin of each individual
  x = matrix(nrow = n, ncol=R) #used to store genotypes
  
  for(i in 1:n){
    z[i] = sample(1:2, size=1, prob=c(w, 1-w))
    x[i,] = r_haploid_genotypes(1,P[z[i],])
  }
  return(list(x=x,z=z))
}</code></pre>
<p>KS - dissecting the function</p>
<pre class="r"><code>n=20
p = seq(0.1,0.9,length=9)
P = rbind(p,1-p)

R = ncol(P) # number of loci
R</code></pre>
<pre><code>[1] 9</code></pre>
<pre class="r"><code>z = rep(0,n) # used to store population of origin of each individual
z</code></pre>
<pre><code> [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</code></pre>
<pre class="r"><code>x = matrix(nrow = n, ncol=R) #used to store genotypes
x</code></pre>
<pre><code>      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9]
 [1,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
 [2,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
 [3,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
 [4,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
 [5,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
 [6,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
 [7,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
 [8,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
 [9,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[10,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[11,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[12,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[13,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[14,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[15,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[16,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[17,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[18,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[19,]   NA   NA   NA   NA   NA   NA   NA   NA   NA
[20,]   NA   NA   NA   NA   NA   NA   NA   NA   NA</code></pre>
<pre class="r"><code>w=0.5
for(i in 1:n){
  z[i] = sample(1:2, size=1, prob=c(w, 1-w))
  print(z[i])
  x[i,] = r_haploid_genotypes(1,P[z[i],])
  print(P[z[i],])
  print(x[i,])
}</code></pre>
<pre><code>[1] 1
[1] 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9
[1] 0 0 0 0 0 1 1 0 0
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 1 0 1 0 0 0 0 0 0
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 0 1 1 1 1 0 0 0 0
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 1 1 1 1 1 0 0 1 1
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 0 1 1 1 0 0 0 0 1
[1] 1
[1] 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9
[1] 0 0 0 0 0 1 0 1 1
[1] 1
[1] 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9
[1] 0 0 0 0 1 1 0 0 1
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 1 1 1 1 1 0 0 0 0
[1] 1
[1] 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9
[1] 0 0 1 0 1 1 1 1 1
[1] 1
[1] 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9
[1] 0 0 0 1 1 0 1 1 1
[1] 1
[1] 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9
[1] 0 0 0 0 0 1 1 1 1
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 1 1 0 0 1 1 0 0 0
[1] 1
[1] 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9
[1] 0 0 1 1 0 1 1 1 1
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 1 1 1 1 1 1 0 0 0
[1] 1
[1] 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9
[1] 0 0 1 0 0 0 1 1 0
[1] 1
[1] 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9
[1] 0 0 1 0 0 0 1 1 0
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 1 1 1 1 0 1 0 0 0
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 0 1 1 0 1 1 0 0 0
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 1 0 1 1 1 0 0 0 0
[1] 2
[1] 0.9 0.8 0.7 0.6 0.5 0.4 0.3 0.2 0.1
[1] 1 0 1 1 1 1 0 0 1</code></pre>
<div id="example-1" class="section level3">
<h3>Example</h3>
<p>Here we sample 20 individuals: in one population the frequencies are as above, whereas in the second population they are reversed (<code>1-p</code>)…again not at all realistic but to help illustrate an idea.</p>
<pre class="r"><code>set.seed(123)
P = rbind(p,1-p)
print(P)</code></pre>
<pre><code>  [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9]
p  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  0.9
   0.9  0.8  0.7  0.6  0.5  0.4  0.3  0.2  0.1</code></pre>
<pre class="r"><code>sim = r_simplemix(n=20, P)
sim</code></pre>
<pre><code>$x
      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9]
 [1,]    1    1    0    0    0    0    1    0    0
 [2,]    0    0    0    0    1    1    1    1    0
 [3,]    0    0    1    1    1    1    1    1    1
 [4,]    1    0    1    0    0    0    1    1    1
 [5,]    1    1    1    1    0    0    0    0    0
 [6,]    1    1    1    1    0    0    1    1    0
 [7,]    0    0    0    1    0    0    0    1    1
 [8,]    0    0    0    0    0    1    1    1    1
 [9,]    1    1    0    1    0    1    1    1    0
[10,]    1    1    1    1    0    1    0    0    0
[11,]    0    0    1    0    1    0    1    1    1
[12,]    0    0    1    1    0    1    0    1    1
[13,]    0    0    0    0    1    1    1    1    1
[14,]    0    0    1    0    1    0    0    0    1
[15,]    1    1    1    0    0    0    0    0    0
[16,]    0    0    0    0    0    1    1    1    1
[17,]    0    0    0    0    1    0    1    1    1
[18,]    0    1    1    1    1    1    1    0    1
[19,]    0    0    0    0    0    1    1    0    0
[20,]    1    0    1    0    0    0    0    0    0

$z
 [1] 2 1 1 1 2 2 1 1 2 2 1 1 1 1 2 1 1 1 1 2</code></pre>
<pre class="r"><code>sim_w = r_simplemix_w(n=20, P, w=0.1)
sim_w</code></pre>
<pre><code>$x
      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9]
 [1,]    0    1    1    1    1    0    0    0    0
 [2,]    1    1    1    1    1    0    0    1    1
 [3,]    0    1    1    1    0    0    0    0    1
 [4,]    1    1    1    1    0    0    1    0    0
 [5,]    1    1    1    1    1    0    1    1    0
 [6,]    1    1    1    1    1    0    0    0    0
 [7,]    0    0    1    0    1    1    1    1    1
 [8,]    1    1    1    0    1    1    0    0    0
 [9,]    1    1    1    1    0    0    0    0    0
[10,]    1    1    0    0    1    1    0    0    0
[11,]    1    1    0    0    0    0    0    0    0
[12,]    1    1    1    1    1    1    0    0    0
[13,]    1    1    0    1    0    1    0    0    1
[14,]    1    1    0    1    0    1    0    0    1
[15,]    1    1    1    1    0    1    0    0    0
[16,]    0    1    1    0    1    1    0    0    0
[17,]    1    0    1    1    1    0    0    0    0
[18,]    1    0    1    1    1    1    0    0    1
[19,]    1    1    0    1    1    0    0    0    0
[20,]    0    0    1    0    0    1    1    1    0

$z
 [1] 2 2 2 2 2 2 1 2 2 2 2 2 2 2 2 2 2 2 2 1</code></pre>
<p>Here are the results of the simulation (first column is the population of origin; the remaining columns are the genotypes). If you look carefully you should see that individuals from population 1 tend to have more <code>1</code> alleles in the later loci, whereas individuals from population <code>2</code> tend to have more <code>1</code> alleles in the earlier loci. This is because of the way that the allele frequencies were set up in the two populations. (Of course in real data the differences between different populations will not usually show patterns like this! I chose the patterns so you can see them by eye, not to be realistic.)</p>
<pre class="r"><code>print(cbind(z=sim$z,sim$x))</code></pre>
<pre><code>      z                  
 [1,] 2 1 1 0 0 0 0 1 0 0
 [2,] 1 0 0 0 0 1 1 1 1 0
 [3,] 1 0 0 1 1 1 1 1 1 1
 [4,] 1 1 0 1 0 0 0 1 1 1
 [5,] 2 1 1 1 1 0 0 0 0 0
 [6,] 2 1 1 1 1 0 0 1 1 0
 [7,] 1 0 0 0 1 0 0 0 1 1
 [8,] 1 0 0 0 0 0 1 1 1 1
 [9,] 2 1 1 0 1 0 1 1 1 0
[10,] 2 1 1 1 1 0 1 0 0 0
[11,] 1 0 0 1 0 1 0 1 1 1
[12,] 1 0 0 1 1 0 1 0 1 1
[13,] 1 0 0 0 0 1 1 1 1 1
[14,] 1 0 0 1 0 1 0 0 0 1
[15,] 2 1 1 1 0 0 0 0 0 0
[16,] 1 0 0 0 0 0 1 1 1 1
[17,] 1 0 0 0 0 1 0 1 1 1
[18,] 1 0 1 1 1 1 1 1 0 1
[19,] 1 0 0 0 0 0 1 1 0 0
[20,] 2 1 0 1 0 0 0 0 0 0</code></pre>
</div>
</div>
<div id="inference-problems" class="section level2">
<h2>Inference problems</h2>
<p>These types of data can help motivate a number of statistical inference problems.</p>
<ol style="list-style-type: decimal">
<li><p>Given the allele frequencies <code>P</code> and the genotypes <code>x</code>, how might you infer the populations of origin <code>z</code>? In the genetic literature this is sometimes called the “assignment problem”; more generally, it is an example of a “classification problem”.</p></li>
<li><p>Given the populations of origin <code>z</code> and the genotypes <code>x</code>, how might you infer the population allele frequencies <code>P</code>?</p></li>
<li><p>Given just the genotypes <code>x</code> how might you infer both <code>z</code> and <code>P</code>? (This is an example of a “clustering” problem).</p></li>
</ol>
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<p>Here are some things you might like to try:</p>
<ol style="list-style-type: decimal">
<li><p>Modify the <code>r_simplemix</code> code to allow the mixture proportions to be specified, rather than fixed at (0.5,0.5). You could do this by adding a parameter <code>w</code> to the function that specifies the proportions to use (<code>w</code> for “weights”).</p></li>
<li><p>Write a function, <code>posterior_prob_assignment=function(x,P,w)</code>, to compute the posterior probability that each individual came from each population, given the genotypes <code>x</code>, the allele frequencies <code>P</code> and the mixture proportions <code>w</code>. Apply your function to the simulated data.</p></li>
<li><p>Write a function <code>posterior_param_allele_frequencies=function(x,z,a)</code> to compute the parameters of the beta posterior in each population at each locus. Here <code>a</code> is a vector of length 2 giving the parameters of the beta prior for <code>P[k,r]</code>. That is, the prior is <code>P[k,r]</code> <span class="math inline">\(\sim\)</span> Beta(<code>a[1]</code>,<code>a[2]</code>). Because the Beta distribution has two parameters, there will be 2 parameters for each locus and each population. So the output of your function should be a 2 x ‘K’ x <code>R</code> array (where ‘K=2’ because we have a mixture of 2 populations).</p></li>
</ol>
</div>
<div id="answer-templates" class="section level2">
<h2>Answer templates</h2>
<p>Here are some answer templates for Exercises 2 and 3. You will need to fill in the “…” (Note: these chunks have the option set <code>eval=FALSE</code> so that this document compiles; you will have to remove these if you want to run these chunks on compilation.)</p>
<div id="exercise-2" class="section level3">
<h3>Exercise 2</h3>
<pre class="r"><code>#&#39; Compute the likelihood for allele frequences p given genotype data x
#&#39; @param p an R vector of allele frequencies
#&#39; @param x an R vector of genotypes
likelihood = function(p,x){
  return(prod(p^x*(1-p)^(1-x)))
}

#&#39; normalize a vector, x, so it&#39;s elements sum to 1
#&#39; @param x a vector
normalize = function(x){x/sum(x)}

#&#39; Compute posterior assignment probabilities
#&#39; @param x an n by R matrix of genotypes
#&#39; @param P a 2 by R matrix of allele frequencies
#&#39; @param w a 2-vector of prior probabilities
posterior_prob_assignment=function(x,P,w){
  n = nrow(x)
  K = length(w)
  post = matrix(nrow=n, ncol=K) # to store the posterior probabilities
  lik = rep(0,K) # a vector to store the likelihoods
  for(i in 1:n){
    for(k in 1:K){
      lik[k] = likelihood(P[k,], x[i,]) * w[k]
      # print(lik[k])
    }
    # print(lik)
    # print(normalize(lik))
    post[i,] = normalize(lik)
    # print(post[i,])
  }
  return(post)
}

posterior_prob_assignment=function(x,P,w){
  n = nrow(x)
  K = length(w)
  post = matrix(nrow=n, ncol=K) # to store the posterior probabilities
  lik = rep(0,K) # a vector to store the likelihoods
  for(i in 1:n){
    for(k in 1:K){
      lik[k] = likelihood(P[k,],x[i,])
    }
    post[i,] = normalize(lik*w)
  }
  return(post)
}
x
P
i = 1
likelihood(P, x[i,])

w = c(0.1, 0.9)
k=1
w[k]
w[2]
n = nrow(x)
K = length(w)
post = matrix(nrow=n, ncol=K) # to store the posterior probabilities
lik = rep(0,K) # a vector to store the likelihoods

for(i in 1:n){
  for(k in 1:K){
    lik[k] = likelihood(P, x[i,]) * w[k]
    print(lik[k])
  }
  # print(lik)
  print(sum(lik))
  print(lik / sum(lik))
  print(normalize(lik))
  # post[i,] = normalize(lik)
  # print(post[i,])
}</code></pre>
<p>Here we can run this on the example data, and compare the posterior probabilities with the true labels.</p>
<pre class="r"><code>set.seed(123)
sim = r_simplemix(n=20,P)
posterior = posterior_prob_assignment(sim$x,P,w=c(0.5,0.5))
plot(sim$z,posterior[,2],xlab=&quot;true population&quot;, ylab=&quot;posterior prob for population 2&quot;)</code></pre>
</div>
<div id="exercise-3" class="section level3">
<h3>Exercise 3</h3>
<pre class="r"><code>#&#39; Compute the posterior parameters for allele frequencies given genotypes 
#&#39; and population labels
#&#39; @param x an n by R matrix of genotypes
#&#39; @param z an n vector of population assignments
#&#39; @param a a 2-vector of prior parameters
posterior_param_allele_frequencies=function(x,z,a){
 K = 2
 R = ncol(x)
 post_param = array(dim=c(2,K,R))
 for(k in 1:K){
   for(r in 1:R){
     ...
   }
 }
 return(post_param)
}</code></pre>
<p>Here we can run this on the example data:</p>
<pre class="r"><code>set.seed(123)
sim = r_simplemix(n=20,P)
posterior_param = posterior_param_allele_frequencies(sim$x,sim$z,a = c(1,1))
posterior_param[,1,] # these should be the posterior parameters for population 1
posterior_param[,2,] # these should be the posterior parameters for population 2</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.3 (2020-10-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252 
[2] LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5        whisker_0.4       knitr_1.33        magrittr_2.0.1   
 [5] workflowr_1.6.2   R6_2.5.0          rlang_0.4.10      fansi_0.4.2      
 [9] stringr_1.4.0     tools_4.0.3       xfun_0.22         utf8_1.1.4       
[13] git2r_0.28.0      jquerylib_0.1.4   htmltools_0.5.1.1 ellipsis_0.3.2   
[17] rprojroot_2.0.2   yaml_2.2.1        digest_0.6.27     tibble_3.0.4     
[21] lifecycle_1.0.0   crayon_1.4.1      later_1.2.0       sass_0.4.0       
[25] vctrs_0.3.8       fs_1.5.0          promises_1.2.0.1  glue_1.4.2       
[29] evaluate_0.14     rmarkdown_2.9     stringi_1.5.3     compiler_4.0.3   
[33] bslib_0.2.5.1     pillar_1.6.0      jsonlite_1.7.2    httpuv_1.6.0     
[37] pkgconfig_2.0.3  </code></pre>
</div>
</div>
</div>

<hr>
<p>
    This site was created with <a href="http://rmarkdown.rstudio.com">R Markdown</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
