<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Kiseok Lee" />

<meta name="date" content="2022-01-22" />

<title>HW2_Bayesian_inference_mixture_models</title>

<script src="site_libs/header-attrs-2.9/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">fiveMinuteStats</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephens999/fiveMinuteStats">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">HW2_Bayesian_inference_mixture_models</h1>
<h4 class="author">Kiseok Lee</h4>
<h4 class="date">2022-01-22</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-01-23
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>fiveMinuteStats/analysis/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed12345code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(12345)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed12345code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(12345)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomKiseokUchicagofiveMinuteStatstreeb97512b028b7476ebda04351ab47e2e749292cf4targetblankb97512ba"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/KiseokUchicago/fiveMinuteStats/tree/b97512b028b7476ebda04351ab47e2e749292cf4" target="_blank">b97512b</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomKiseokUchicagofiveMinuteStatstreeb97512b028b7476ebda04351ab47e2e749292cf4targetblankb97512ba" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/KiseokUchicago/fiveMinuteStats/tree/b97512b028b7476ebda04351ab47e2e749292cf4" target="_blank">b97512b</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  texput.log

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/HW2_Bayesian_inference_mixture_models.Rmd</code>) and HTML (<code>docs/HW2_Bayesian_inference_mixture_models.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/KiseokUchicago/fiveMinuteStats/blob/b97512b028b7476ebda04351ab47e2e749292cf4/analysis/HW2_Bayesian_inference_mixture_models.Rmd" target="_blank">b97512b</a>
</td>
<td>
Kiseok Lee
</td>
<td>
2022-01-23
</td>
<td>
wflow_publish(“analysis/HW2_Bayesian_inference_mixture_models.Rmd”)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/KiseokUchicago/fiveMinuteStats/46cfa0aa7cbd3b148e7f495e96b09092d9ba6c66/docs/HW2_Bayesian_inference_mixture_models.html" target="_blank">46cfa0a</a>
</td>
<td>
Kiseok Lee
</td>
<td>
2022-01-22
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/KiseokUchicago/fiveMinuteStats/blob/a2deeb7635f8952fcb50bde09d978341bea1bd95/analysis/HW2_Bayesian_inference_mixture_models.Rmd" target="_blank">a2deeb7</a>
</td>
<td>
Kiseok Lee
</td>
<td>
2022-01-22
</td>
<td>
wflow_publish(“analysis/HW2_Bayesian_inference_mixture_models.Rmd”)
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="hw2-bayesian-inference-mixture-models" class="section level2">
<h2>HW2 Bayesian inference; mixture models</h2>
<p>Name: <strong>Kiseok Lee</strong><br />
Date: 1/22/22<br />
Class: <strong>HGEN 486 Computational Biology</strong></p>
</div>
<div id="a-2-calculate-point-estimate-and-interval-estimate" class="section level2">
<h2>A-(2) Calculate point estimate and interval estimate</h2>
<pre class="r"><code># plot gamma distribution
shape = 12
scale = 1/9
# define x-values
x &lt;- seq(0, 3, by = 0.01)
# calculate gamma density for each
# x-value
y &lt;- dgamma(x, shape = 12, scale = 1/9)

# create density plot
plot(x, y, type = &quot;l&quot;, col = &quot;blue&quot;, xlab = &quot;x&quot;,
    ylab = &quot;density&quot;, main = &quot;Gamma density function (k=12, theta=1/9)&quot;)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-2-1.png" width="1056" style="display: block; margin: auto;" /></p>
<pre class="r"><code>require(pracma)
trapz(x, y)  # area sum to 1</code></pre>
<pre><code>[1] 0.9995737</code></pre>
<pre class="r"><code># posterior median
qgamma(0.5, shape = 12, scale = 1/9)</code></pre>
<pre><code>[1] 1.296485</code></pre>
<pre class="r"><code># 90% credible interval
lower = qgamma(0.05, shape = 12, scale = 1/9)
lower</code></pre>
<pre><code>[1] 0.7693569</code></pre>
<pre class="r"><code>upper = qgamma(0.95, shape = 12, scale = 1/9)
upper</code></pre>
<pre><code>[1] 2.023057</code></pre>
<pre class="r"><code>plot(x, y, type = &quot;l&quot;, col = &quot;blue&quot;, xlab = &quot;x&quot;,
    ylab = &quot;density&quot;, main = &quot;90% credibility interval&quot;)
abline(v = lower, col = &quot;red&quot;)
abline(v = upper, col = &quot;red&quot;)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-2-2.png" width="1056" style="display: block; margin: auto;" /></p>
</div>
<div id="b.-mixture-simulation" class="section level2">
<h2>B. Mixture simulation</h2>
</div>
<div id="b-a-write-an-r-function-to-simulate-genetic-data" class="section level2">
<h2>B-a) Write an R function to simulate genetic data</h2>
<p>Write an R function to simulate genetic data on (haploid) tusks from a mixture distribution: that is, sampled from a population in which each tusk comes from one of K populations.<br />
</p>
<p>reference1: <a href="https://stephens999.github.io/fiveMinuteStats/mixture_models_01.html" class="uri">https://stephens999.github.io/fiveMinuteStats/mixture_models_01.html</a><br />
reference2: <a href="https://stephens999.github.io/fiveMinuteStats/likelihood_ratio_simple_models.html" class="uri">https://stephens999.github.io/fiveMinuteStats/likelihood_ratio_simple_models.html</a><br />
</p>
<p>First let’s build the components of the function.</p>
<pre class="r"><code># Input
#&#39; @param N  number of tusks to simulate
#&#39; @param vec_w vector of mixture proportions, a non-negative vector of lenght K that sums to 1.
#&#39; @param M  number of markers
#&#39; @param mat_F K by M matrix, F, of allele frequencies

# Output
#&#39; @param vec_z  # Z vector of component memberships
#&#39; @param mat_X  # N by M matrix of simulated data

# 2 step process


N = 1000  # number of tusks to simulate
M = 6  # number of markers
vec_w = c(0.25, 0.75)  # c(savannah, Forest population)

mat_F = rbind(c(0.4, 0.12, 0.21, 0.12, 0.02,
    0.32), c(0.8, 0.2, 0.11, 0.17, 0.23,
    0.25))  # K by M matrix, F, of allele frequencies
mat_F[1, ]  # savannah allele frequencies</code></pre>
<pre><code>[1] 0.40 0.12 0.21 0.12 0.02 0.32</code></pre>
<pre class="r"><code>mat_F[2, ]  # forest allele frequencies</code></pre>
<pre><code>[1] 0.80 0.20 0.11 0.17 0.23 0.25</code></pre>
<pre class="r"><code># output setting
vec_z = rep(-1, N)  # Z vector of component memberships
length(vec_z)</code></pre>
<pre><code>[1] 1000</code></pre>
<pre class="r"><code>mat_X = matrix(NA, nrow = N, ncol = 6)  # N by M matrix of simulated data
dim(mat_X)</code></pre>
<pre><code>[1] 1000    6</code></pre>
<pre class="r"><code># simulating data
for (i in 1:N) {
    # print(paste0(&#39;i: &#39;,i)) generate
    # latent variable Z in the first
    # step
    if (runif(1) &lt; 0.25) {
        vec_z[i] = 0  # z = 0 is savannah, z = 1, is forest
        # print(vec_z[i])
        mat_X[i, ] = rbinom(M, rep(1, M),
            mat_F[1, ])  # simulate genotype
        # print(mat_X[i,])
    } else {
        vec_z[i] = 1  # z = 0 is savannah, z = 1, is forest
        # print(vec_z[i])
        mat_X[i, ] = rbinom(M, rep(1, M),
            mat_F[2, ])  # simulate genotype
        # print(mat_X[i,])
    }
}

# confirm that I simulated right
mean(vec_z)</code></pre>
<pre><code>[1] 0.754</code></pre>
<pre class="r"><code>colMeans(mat_X[vec_z == 0, ]) - mat_F[1,
    ]  # similar</code></pre>
<pre><code>[1] -0.034146341 -0.006178862 -0.006747967 -0.010243902  0.012520325
[6] -0.039512195</code></pre>
<pre class="r"><code>colMeans(mat_X[vec_z == 1, ]) - mat_F[2,
    ]  # similar</code></pre>
<pre><code>[1]  0.006366048 -0.003713528 -0.005225464 -0.022785146  0.011379310
[6]  0.020557029</code></pre>
<p>Wrap it into a function</p>
<pre class="r"><code>## Wrapping into a function

# Input
#&#39; @param N  number of tusks to simulate
#&#39; @param vec_w vector of mixture proportions, a non-negative vector of lenght K that sums to 1.
#&#39; @param M  number of markers
#&#39; @param mat_F K by M matrix, F, of allele frequencies

# Output
#&#39; @param vec_z  # Z vector of component memberships
#&#39; @param mat_X  # N by M matrix of simulated data

simulate_data &lt;- function(N, vec_w, M, mat_F) {
    # output setting
    vec_z = rep(-1, N)  # Z vector of component memberships
    length(vec_z)
    mat_X = matrix(NA, nrow = N, ncol = 6)  # N by M matrix of simulated data
    dim(mat_X)

    # simulating data
    for (i in 1:N) {
        # print(paste0(&#39;i: &#39;,i))
        # generate latent variable Z in
        # the first step
        if (runif(1) &lt; 0.25) {
            vec_z[i] = 0  # z = 0 is savannah, z = 1, is forest
            # print(vec_z[i])
            mat_X[i, ] = rbinom(M, rep(1,
                M), mat_F[1, ])  # simulate genotype
            # print(mat_X[i,])
        } else {
            vec_z[i] = 1  # z = 0 is savannah, z = 1, is forest
            # print(vec_z[i])
            mat_X[i, ] = rbinom(M, rep(1,
                M), mat_F[2, ])  # simulate genotype
            # print(mat_X[i,])
        }
    }
    out_list = list(Component_membership = vec_z,
        Simulated_data = mat_X)
    return(out_list)

}

# example
N = 1000  # number of tusks to simulate
M = 6  # number of markers
vec_w = c(0.25, 0.75)  # c(savannah, Forest population)

mat_F = rbind(c(0.4, 0.12, 0.21, 0.12, 0.02,
    0.32), c(0.8, 0.2, 0.11, 0.17, 0.23,
    0.25))  # K by M matrix, F, of allele frequencies

output_list &lt;- simulate_data(N = N, vec_w = vec_w,
    M = M, mat_F = mat_F)

output_list$Component_membership</code></pre>
<pre><code>   [1] 1 1 1 1 1 1 1 1 0 1 1 1 1 1 0 1 0 1 0 1 1 1 1 1 1 1 1 0 1 0 1 0 1 0 1 0 0
  [38] 0 1 0 1 1 0 1 1 1 1 1 0 1 1 1 0 0 1 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 0
  [75] 0 1 1 1 1 1 0 0 0 0 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 0 1 0 1 1 1
 [112] 1 1 1 1 1 0 1 0 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1
 [149] 0 1 1 1 1 0 1 0 1 0 1 1 1 1 0 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 0 1 1 1 1
 [186] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 0 1 0 1 1 0 1 1 1 1 0 1 1 1
 [223] 1 0 0 1 1 1 1 0 0 1 1 0 1 1 1 0 0 1 1 1 1 1 1 1 1 0 1 0 1 1 1 1 0 0 0 1 1
 [260] 1 1 1 0 0 0 0 1 1 1 0 0 1 1 1 1 1 0 1 1 1 1 0 1 1 1 1 0 0 1 1 1 0 1 1 1 0
 [297] 1 1 1 0 1 1 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1
 [334] 1 1 0 0 1 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 0 1 1 0 0 1 1 1 1
 [371] 1 1 0 1 0 0 1 1 0 1 0 1 1 0 1 1 1 1 0 1 0 1 1 1 0 1 1 0 0 1 1 1 1 1 1 1 0
 [408] 1 1 1 1 1 1 1 0 1 1 1 0 1 1 1 1 1 0 1 0 0 1 1 1 1 0 0 1 0 1 1 1 1 1 1 1 1
 [445] 0 1 0 1 1 1 0 0 1 1 0 0 1 1 1 1 1 0 1 1 0 1 1 0 1 1 1 1 0 1 0 1 1 0 0 1 0
 [482] 1 1 1 1 0 1 1 0 0 1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1
 [519] 1 1 1 1 1 1 1 1 0 1 0 1 1 1 0 1 0 1 0 0 0 1 0 0 1 1 1 1 1 1 1 1 1 1 0 1 1
 [556] 0 1 1 1 1 0 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 0 1 1 1 1 1 1 1 1 1 1
 [593] 1 1 1 0 0 1 1 1 1 1 1 1 1 0 1 0 0 1 0 0 0 0 1 1 1 0 1 1 0 1 0 0 0 1 1 1 0
 [630] 1 1 0 0 1 0 0 1 1 1 1 1 0 1 1 1 1 1 1 0 1 1 0 1 0 1 1 1 1 1 0 1 1 1 1 1 1
 [667] 1 0 0 1 1 1 1 1 1 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0
 [704] 1 1 1 1 1 1 1 1 1 0 1 0 1 1 0 1 1 1 1 1 0 1 0 1 0 0 0 0 1 1 1 1 1 1 1 1 1
 [741] 0 0 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 1 1 1 1 0 1 1 1 1 1 0 0 1 1 1
 [778] 1 1 1 0 1 1 0 1 0 1 0 0 0 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1
 [815] 1 1 1 1 1 0 1 1 0 1 0 1 1 1 1 1 1 0 1 1 1 1 1 1 0 0 0 0 1 1 0 1 0 1 0 1 0
 [852] 1 1 1 1 1 1 0 1 1 0 1 0 1 1 1 0 1 0 1 1 1 1 0 1 0 1 1 1 0 1 1 0 0 1 0 1 1
 [889] 1 0 1 1 0 1 1 0 1 1 0 0 0 1 1 0 1 1 1 1 1 1 1 1 0 0 1 0 1 1 0 1 0 1 1 0 1
 [926] 1 0 0 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 1 1 1 1 1 0 0 1 0 1 0 1 0 1 0 1 1 1
 [963] 0 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 1 1 0
[1000] 1</code></pre>
<pre class="r"><code>head(output_list$Simulated_data)</code></pre>
<pre><code>     [,1] [,2] [,3] [,4] [,5] [,6]
[1,]    1    0    0    0    0    1
[2,]    0    0    0    1    0    1
[3,]    1    0    0    0    1    0
[4,]    1    0    0    0    0    0
[5,]    1    0    0    0    1    0
[6,]    0    0    0    1    1    0</code></pre>
</div>
<div id="b-b-write-an-r-function-to-compute-the-log-likelihood" class="section level2">
<h2>B-b) Write an R function to compute the log-likelihood</h2>
<p>Write an R function to compute the log-likelihood l(w) for a data set simulated from the mixture model you have just implemented.<br />
</p>
<p>Construct the elements of the function</p>
<pre class="r"><code>X &lt;- output_list$Simulated_data

# Input
#&#39; @param X  simulated data matrix
#&#39; @param vec_w vector of mixture proportions, a non-negative vector of lenght K that sums to 1.
#&#39; @param mat_F K by M matrix, F, of allele frequencies

# likelihood under model
likelihood_under_model = function(p, x) {
    return(prod(p^x * (1 - p)^(1 - x)))
}

# loop thru whole dataset and multiply
# to get likelihood
log_likelihood = 0
for (i in 1:dim(X)[1]) {
    # get likelihood of one simulation
    likelihood_each_simulation = vec_w[1] *
        likelihood_under_model(mat_F[1, ],
            X[i, ]) + vec_w[2] * likelihood_under_model(mat_F[2,
        ], X[i, ])
    # print(likelihood_each_simulation)
    log_likelihood = log_likelihood + log(likelihood_each_simulation)
    # print(likelihood)
}


# function
log_lik &lt;- function(w_1, X, mat_F) {
    log_likelihood = 0
    for (i in 1:dim(X)[1]) {
        # get likelihood of one
        # simulation
        likelihood_each_simulation = w_1 *
            likelihood_under_model(mat_F[1,
                ], X[i, ]) + (1 - w_1) *
            likelihood_under_model(mat_F[2,
                ], X[i, ])
        # print(likelihood_each_simulation)
        log_likelihood = log_likelihood +
            log(likelihood_each_simulation)
        # print(likelihood)
    }
    return(log_likelihood)
}</code></pre>
</div>
<div id="b-c-perform-a-preliminary-test-of-your-likelihood-and-simulation-programs" class="section level2">
<h2>B-c) Perform a preliminary test of your likelihood and simulation programs</h2>
<pre class="r"><code># example
N = 1000  # number of tusks to simulate
M = 6  # number of markers
vec_w = c(0.25, 0.75)  # c(savannah, Forest population)

mat_F = rbind(c(0.4, 0.12, 0.21, 0.12, 0.02,
    0.32), c(0.8, 0.2, 0.11, 0.17, 0.23,
    0.25))  # K by M matrix, F, of allele frequencies

output_list &lt;- simulate_data(N = N, vec_w = vec_w,
    M = M, mat_F = mat_F)

X &lt;- output_list$Simulated_data

vec_w1 = seq(0, 1, length = 100)
vec_loglik = rep(-1, 100)
for (i in 1:length(vec_w1)) {
    vec_loglik[i] &lt;- log_lik(vec_w1[i], X,
        mat_F)
}

plot(vec_w1, vec_loglik, type = &quot;l&quot;, main = &quot;log-likelihood&quot;,
    xlab = &quot;w_1&quot;, ylab = &quot;Log likelihood&quot;)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-6-1.png" width="1056" style="display: block; margin: auto;" /></p>
<pre class="r"><code># zoom in
plot(vec_w1, vec_loglik - max(vec_loglik),
    ylim = c(-10, 0), type = &quot;l&quot;, main = &quot;log-likelihood&quot;,
    xlab = &quot;w_1&quot;, ylab = &quot;Log likelihood&quot;)
abline(v = 0.25, col = &quot;red&quot;)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-6-2.png" width="1056" style="display: block; margin: auto;" /></p>
<p>It is not exactly maximized at 0.25 but still close enough.</p>
</div>
<div id="c.-em-algorithm-for-estimating-mixture-component-proportions" class="section level2">
<h2>C. EM algorithm for estimating mixture component proportions</h2>
<p>Reread <a href="https://stephens999.github.io/fiveMinuteStats/em_algorithm_01.html" class="uri">https://stephens999.github.io/fiveMinuteStats/em_algorithm_01.html</a><br />
</p>
</div>
<div id="c-a-write-a-function-to-estimate-mixture-proportions-in-a-generalized-form" class="section level2">
<h2>C-a) Write a function to estimate mixture proportions in a generalized form</h2>
<pre class="r"><code># Input
#&#39; @param L  likelihood matrix
#&#39; @param k  number of components
#&#39; @param pi_init initial values for pi, length of k

general_em &lt;- function(L, pi_init, niter = 100) {
    n = dim(L)[1]  # number of data
    k = dim(L)[2]  # number of components or populations
    mat_pi = matrix(nrow = niter + 1, ncol = k)
    mat_pi[1, ] = pi_init  # set initial pi 
    # print(head(mat_pi))
    w = matrix(nrow = n, ncol = k)

    for (i in 1:niter) {
        # print(i) calculate likelihood
        # * prior
        for (j in 1:n) {
            w[j, ] = mat_pi[i, ] * L[j, ]
        }
        w_norm = t(apply(w, 1, function(x) x/sum(x)))  # calculate posterior by normalizing by row
        # get pi vector by averaging
        # posterior for each population
        mat_pi[i + 1, ] &lt;- apply(w_norm,
            2, mean)
        # print(mat_pi[i+1,])
    }
    return(mat_pi)
}

# Let&#39;s see if we did a good job by
# comparing with simple_em

# example from
# https://stephens999.github.io/fiveMinuteStats/em_algorithm_01.html

n = 1000  # number of samples
x = rep(0, n)  # to store the samples
shape = c(2, 2)  # shapes of the two components
scale = c(0.5, 1)  # scales of the two components
for (i in 1:n) {
    if (runif(1) &lt; 0.7)
        z = 1 else z = 2
    x[i] = rgamma(1, scale = scale[z], shape = shape[z])
}

mix_loglik = function(pi1, x) {
    sum(log(pi1 * dgamma(x, scale = 0.5,
        shape = 2) + (1 - pi1) * dgamma(x,
        scale = 1, shape = 2)))
}

pi1_vec = seq(0, 1, length = 100)
loglik = rep(0, 100)
for (i in 1:length(pi1_vec)) {
    loglik[i] = mix_loglik(pi1_vec[i], x)
}

niter = 100
pi1.init = 0.5
n = length(x)
pi1 = rep(0, niter + 1)
pi1[1] = pi1.init
L = matrix(nrow = n, ncol = 2)
L[, 1] = dgamma(x, scale = 0.5, shape = 2)
L[, 2] = dgamma(x, scale = 1, shape = 2)

## try with simple_em
simple_em = function(x, pi1.init = 0.5, niter = 100) {
    n = length(x)
    pi1 = rep(0, niter + 1)
    pi1[1] = pi1.init
    L = matrix(nrow = n, ncol = 2)
    L[, 1] = dgamma(x, scale = 0.5, shape = 2)
    L[, 2] = dgamma(x, scale = 1, shape = 2)
    w = matrix(nrow = n, ncol = 2)
    for (iter in 1:niter) {
        w[, 1] = pi1[iter] * L[, 1]
        w[, 2] = (1 - pi1[iter]) * L[, 2]
        for (i in 1:n) {
            # normalize the rows of w
            # to sum to 1
            w[i, ] = w[i, ]/(w[i, 1] + w[i,
                2])
        }
        pi1[iter + 1] = mean(w[, 1])
    }
    return(pi1)

}

pi1.iter = simple_em(x)
plot(pi1.iter, xlab = &quot;iteration&quot;, ylab = &quot;pi1&quot;,
    type = &quot;b&quot;, col = &quot;blue&quot;, main = &quot;simple_em&quot;)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-7-1.png" width="1056" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## try with general_em
pi.iter = general_em(L, c(0.5, 0.5), niter = 100)
plot(pi.iter[, 1], xlab = &quot;iteration&quot;, ylab = &quot;pi1&quot;,
    col = &quot;red&quot;, type = &quot;b&quot;, lty = 1, pch = 19,
    lwd = 1, main = &quot;general_em&quot;)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-7-2.png" width="1056" style="display: block; margin: auto;" /></p>
<pre class="r"><code># legend(&#39;bottomright&#39;,
# legend=c(&#39;simple_em&#39;, &#39;general_em&#39;),
# col=c(&#39;Blue&#39;, &#39;Red&#39;), lty=1, cex=0.8,
# box.lty=0)</code></pre>
</div>
<div id="c-b-apply-your-algorithm-by-applying-it-to-genetic-mixture-data" class="section level2">
<h2>C-b) Apply your algorithm by applying it to genetic mixture data</h2>
<p>Test your algorithm by applying it to genetic mixture data simulated as in B. Do at least 2 simulated data sets, one with more than 2 mixture components. Check that the log-likelihood is increasing each iteration and compare your estimated mixture proportions with the truth. Check whether you get the same answer from different starting points.</p>
<div id="st-simulated-dataset-with-2-mixture-components-savannah-vs-forest-population" class="section level3">
<h3>(1) 1st simulated dataset with 2 mixture components (savannah vs forest population)</h3>
<p>(1-1) Compare the estimated mixture proportions with the truth</p>
<pre class="r"><code># Genetic mixture data simulated in B

# Input
N = 1000  # number of tusks to simulate
M = 6  # number of markers
K = 2  # number of components
mat_F = rbind(c(0.4, 0.12, 0.21, 0.12, 0.02,
    0.32), c(0.8, 0.2, 0.11, 0.17, 0.23,
    0.25))  # K by M matrix, F, of allele frequencies

# simulate data
output_list &lt;- simulate_data(N = N, vec_w = vec_w,
    M = M, mat_F = mat_F)
X &lt;- output_list$Simulated_data

# Now calculated likelihood under each
# model loop thru whole dataset and
# multiply to get likelihood
L = matrix(NA, nrow = N, ncol = K)
for (i in 1:dim(X)[1]) {
    # get likelihood of one simulation
    for (j in 1:K) {
        # print(j)
        L[i, j] = likelihood_under_model(mat_F[j,
            ], X[i, ])  # calculate likelihood under each model
        # print(L[i,])
    }
}

dim(L)</code></pre>
<pre><code>[1] 1000    2</code></pre>
<pre class="r"><code>## EM
pi_iteration &lt;- general_em(L, c(0.5, 0.5),
    niter = 100)

# plot and compare with true component
# proportion value
plot(pi_iteration[, 1], xlab = &quot;iteration&quot;,
    ylab = &quot;Component proportion&quot;, col = &quot;red&quot;,
    type = &quot;b&quot;, lty = 1, pch = 19, lwd = 1,
    main = &quot;EM on genetic mixture data&quot;,
    ylim = c(0, 1))
lines(pi_iteration[, 2], col = &quot;blue&quot;, type = &quot;b&quot;,
    lty = 1, pch = 19, lwd = 1)
abline(h = 0.25, col = &quot;red&quot;, lty = 3, lwd = 2)
abline(h = 0.75, col = &quot;blue&quot;, lty = 3, lwd = 2)
legend(&quot;bottomright&quot;, legend = c(&quot;Savannah&quot;,
    &quot;Forest&quot;, &quot;Truth&quot;), col = c(&quot;Red&quot;, &quot;Blue&quot;,
    &quot;Black&quot;), lty = c(1, 1, 3), cex = 0.8,
    box.lty = 0)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-8-1.png" width="1056" style="display: block; margin: auto;" /></p>
<pre class="r"><code>tail(pi_iteration, 1)  # this is the EM estimated proportion value</code></pre>
<pre><code>            [,1]      [,2]
[101,] 0.2677262 0.7322738</code></pre>
<p>Truth was pi1 = 0.25, pi2 = 0.75. The EM estimate is close to 0.25 but not exactly 0.25.</p>
<p>(1-2) Compare with different starting point</p>
<pre class="r"><code>## EM
pi_iteration_0.1 &lt;- general_em(L, c(0.1,
    0.9), niter = 100)
pi_iteration_0.2 &lt;- general_em(L, c(0.2,
    0.8), niter = 100)
pi_iteration_0.3 &lt;- general_em(L, c(0.3,
    0.7), niter = 100)
pi_iteration_0.4 &lt;- general_em(L, c(0.4,
    0.6), niter = 100)
pi_iteration_0.5 &lt;- general_em(L, c(0.5,
    0.5), niter = 100)

# plot and compare with true component
# proportion value
plot(pi_iteration_0.1[, 1], xlab = &quot;iteration&quot;,
    ylab = &quot;pi 1&quot;, col = &quot;red&quot;, type = &quot;l&quot;,
    lty = 1, pch = 19, lwd = 2, main = &quot;EM with different initial pi 1&quot;,
    ylim = c(0, 1))
lines(pi_iteration_0.2[, 1], col = &quot;orange&quot;,
    type = &quot;l&quot;, lty = 1, pch = 19, lwd = 2)
lines(pi_iteration_0.3[, 1], col = &quot;green&quot;,
    type = &quot;l&quot;, lty = 1, pch = 19, lwd = 2)
lines(pi_iteration_0.4[, 1], col = &quot;blue&quot;,
    type = &quot;l&quot;, lty = 1, pch = 19, lwd = 2)
lines(pi_iteration_0.5[, 1], col = &quot;purple&quot;,
    type = &quot;l&quot;, lty = 1, pch = 19, lwd = 2)
legend(60, 0.8, legend = c(&quot;init_pi1=0.1&quot;,
    &quot;init_pi1=0.2&quot;, &quot;init_pi1=0.3&quot;, &quot;init_pi1=0.4&quot;,
    &quot;init_pi1=0.5&quot;), col = c(&quot;red&quot;, &quot;orange&quot;,
    &quot;green&quot;, &quot;blue&quot;, &quot;purple&quot;), lty = 1,
    cex = 0.6, box.lty = 0)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-9-1.png" width="1056" style="display: block; margin: auto;" /> Different starting pi_1 all converges to the same pi_1.</p>
<p>(1-3) Check log-likelihood is increasing: True</p>
<pre class="r"><code>## Track log-likelihood changes
niter = 100
vec_ll = rep(Inf, niter)  # log likelihood vector
for (i in 1:dim(pi_iteration)[1]) {
    # print(i)
    log_likelihood = 0
    for (j in 1:dim(L)[1]) {
        # print(j) get likelihood of
        # one simulation
        likelihood_each_simulation = sum(pi_iteration[i,
            ] * L[j, ])
        # print(likelihood_each_simulation)
        log_likelihood = log_likelihood +
            log(likelihood_each_simulation)

    }
    # print(log_likelihood)
    vec_ll[i] = log_likelihood
}

# plot the change of log-likelihood
plot(vec_ll, xlab = &quot;iteration&quot;, ylab = &quot;Log-likelihood&quot;,
    col = &quot;maroon&quot;, type = &quot;b&quot;, lty = 1,
    pch = 19, lwd = 1, main = &quot;Log-likelihood change during EM&quot;)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-10-1.png" width="1056" style="display: block; margin: auto;" /></p>
</div>
<div id="nd-simulated-dataset-with-3-mixture-components-savannah-vs-forest-vs-island-population" class="section level3">
<h3>(2) 2nd simulated dataset with 3 mixture components (savannah vs forest vs island population)</h3>
<p>Adding third component: island tusks</p>
<p>(1-1) Compare the estimated mixture proportions with the truth</p>
<pre class="r"><code># Genetic mixture data with 3
# components

# Input
N = 1000  # number of tusks to simulate
M = 6  # number of markers
K = 3  # number of components
mat_F = rbind(c(0.4, 0.12, 0.21, 0.12, 0.02,
    0.32), c(0.8, 0.2, 0.11, 0.17, 0.23,
    0.25), c(0.2, 0.3, 0.8, 0.9, 0.7, 0.4))  # K by M matrix, F, of allele frequencies
vec_w = c(0.6, 0.3, 0.1)

# simulate data
output_list &lt;- simulate_data(N = N, vec_w = vec_w,
    M = M, mat_F = mat_F)
X &lt;- output_list$Simulated_data
dim(X)</code></pre>
<pre><code>[1] 1000    6</code></pre>
<pre class="r"><code># Now calculated likelihood under each
# model loop thru whole dataset and
# multiply to get likelihood
L = matrix(NA, nrow = N, ncol = K)
dim(L)</code></pre>
<pre><code>[1] 1000    3</code></pre>
<pre class="r"><code>for (i in 1:dim(X)[1]) {
    # get likelihood of one simulation
    for (j in 1:K) {
        # print(j)
        L[i, j] = likelihood_under_model(mat_F[j,
            ], X[i, ])  # calculate likelihood under each model
        # print(L[i,])
    }
}

## EM
pi_iteration &lt;- general_em(L, pi_init = c(0.4,
    0.3, 0.3), niter = 100)

# plot and compare with true component
# proportion value
plot(pi_iteration[, 1], xlab = &quot;iteration&quot;,
    ylab = &quot;Component proportion&quot;, col = &quot;red&quot;,
    type = &quot;b&quot;, lty = 1, pch = 19, lwd = 1,
    main = &quot;EM on genetic mixture data&quot;,
    ylim = c(0, 1))
lines(pi_iteration[, 2], col = &quot;blue&quot;, type = &quot;b&quot;,
    lty = 1, pch = 19, lwd = 1)
lines(pi_iteration[, 3], col = &quot;purple&quot;,
    type = &quot;b&quot;, lty = 1, pch = 19, lwd = 1)
abline(h = 0.6, col = &quot;red&quot;, lty = 3, lwd = 2)
abline(h = 0.3, col = &quot;blue&quot;, lty = 3, lwd = 2)
abline(h = 0.1, col = &quot;purple&quot;, lty = 3,
    lwd = 2)

legend(&quot;bottomright&quot;, legend = c(&quot;Savannah&quot;,
    &quot;Forest&quot;, &quot;Island&quot;, &quot;Truth&quot;), col = c(&quot;Red&quot;,
    &quot;Blue&quot;, &quot;Purple&quot;, &quot;Black&quot;), lty = c(1,
    1, 1, 3), cex = 0.8, box.lty = 0)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-11-1.png" width="1056" style="display: block; margin: auto;" /></p>
<pre class="r"><code>tail(pi_iteration, 1)  # this is the EM estimated proportion value</code></pre>
<pre><code>            [,1]      [,2]         [,3]
[101,] 0.2300216 0.7699784 1.424474e-35</code></pre>
<p>Truth was c(0.6, 0.3, 0.1). The EM estimate is far from the true mixture component proportion.</p>
<p>(1-2) Compare with different starting point</p>
<pre class="r"><code>## EM
pi_iteration_0.1 &lt;- general_em(L, c(0.1,
    0.2, 0.7), niter = 100)
pi_iteration_0.3 &lt;- general_em(L, c(0.3,
    0.4, 0.3), niter = 100)
pi_iteration_0.5 &lt;- general_em(L, c(0.5,
    0.2, 0.3), niter = 100)
pi_iteration_0.6 &lt;- general_em(L, c(0.6,
    0.2, 0.2), niter = 100)
pi_iteration_0.7 &lt;- general_em(L, c(0.7,
    0.1, 0.2), niter = 100)

# plot and compare with true component
# proportion value
plot(pi_iteration_0.1[, 1], xlab = &quot;iteration&quot;,
    ylab = &quot;pi 1&quot;, col = &quot;red&quot;, type = &quot;l&quot;,
    lty = 1, pch = 19, lwd = 2, main = &quot;EM with different initial pi 1&quot;,
    ylim = c(0, 1))
lines(pi_iteration_0.3[, 1], col = &quot;orange&quot;,
    type = &quot;l&quot;, lty = 1, pch = 19, lwd = 2)
lines(pi_iteration_0.5[, 1], col = &quot;green&quot;,
    type = &quot;l&quot;, lty = 1, pch = 19, lwd = 2)
lines(pi_iteration_0.6[, 1], col = &quot;blue&quot;,
    type = &quot;l&quot;, lty = 1, pch = 19, lwd = 2)
lines(pi_iteration_0.7[, 1], col = &quot;purple&quot;,
    type = &quot;l&quot;, lty = 1, pch = 19, lwd = 2)
abline(h = 0.6, col = &quot;black&quot;, lty = 2, pch = 19,
    lwd = 1)
legend(60, 0.9, legend = c(&quot;init_pi1=0.1&quot;,
    &quot;init_pi1=0.3&quot;, &quot;init_pi1=0.5&quot;, &quot;init_pi1=0.6&quot;,
    &quot;init_pi1=0.7&quot;, &quot;truth&quot;), col = c(&quot;red&quot;,
    &quot;orange&quot;, &quot;green&quot;, &quot;blue&quot;, &quot;purple&quot;,
    &quot;black&quot;), lty = c(rep(1, 5), 2), cex = 0.6,
    box.lty = 0)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-12-1.png" width="1056" style="display: block; margin: auto;" /> Different starting pi_1 all converges to the same pi_1. However, the converged pi_1 is far from truth.</p>
<p>(1-3) Check log-likelihood is increasing: True</p>
<pre class="r"><code>## Track log-likelihood changes
niter = 100
vec_ll = rep(Inf, niter)  # log likelihood vector
for (i in 1:dim(pi_iteration)[1]) {
    # print(i)
    log_likelihood = 0
    for (j in 1:dim(L)[1]) {
        # print(j) get likelihood of
        # one simulation
        likelihood_each_simulation = sum(pi_iteration[i,
            ] * L[j, ])
        # print(likelihood_each_simulation)
        log_likelihood = log_likelihood +
            log(likelihood_each_simulation)

    }
    # print(log_likelihood)
    vec_ll[i] = log_likelihood
}

# plot the change of log-likelihood
plot(vec_ll, xlab = &quot;iteration&quot;, ylab = &quot;Log-likelihood&quot;,
    col = &quot;maroon&quot;, type = &quot;b&quot;, lty = 1,
    pch = 19, lwd = 1, main = &quot;Log-likelihood change during EM&quot;)</code></pre>
<p><img src="figure/HW2_Bayesian_inference_mixture_models.Rmd/unnamed-chunk-13-1.png" width="1056" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.0.3 (2020-10-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252 
[2] LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] pracma_2.3.3       rmarkdown_2.9      ggpubr_0.4.0       ggrepel_0.9.1     
 [5] ape_5.5            openxlsx_4.2.3     devtools_2.4.0     usethis_2.0.1     
 [9] gtools_3.8.2       reshape2_1.4.4     readxl_1.3.1       magrittr_2.0.1    
[13] forcats_0.5.1      stringr_1.4.0      purrr_0.3.4        readr_1.4.0       
[17] tidyr_1.1.3        tibble_3.0.4       tidyverse_1.3.1    vegan_2.5-7       
[21] lattice_0.20-41    permute_0.9-5      RColorBrewer_1.1-2 ggplot2_3.3.5     
[25] dplyr_1.0.5        knitr_1.37        

loaded via a namespace (and not attached):
 [1] colorspace_2.0-0  ggsignif_0.6.2    ellipsis_0.3.2    rio_0.5.27       
 [5] rprojroot_2.0.2   fs_1.5.0          rstudioapi_0.13   remotes_2.4.0    
 [9] fansi_0.4.2       lubridate_1.7.10  xml2_1.3.2        splines_4.0.3    
[13] cachem_1.0.4      pkgload_1.2.1     jsonlite_1.7.2    workflowr_1.6.2  
[17] broom_0.7.9       cluster_2.1.0     dbplyr_2.1.1      compiler_4.0.3   
[21] httr_1.4.2        backports_1.2.1   assertthat_0.2.1  Matrix_1.2-18    
[25] fastmap_1.1.0     cli_3.0.1         formatR_1.11      later_1.2.0      
[29] htmltools_0.5.1.1 prettyunits_1.1.1 tools_4.0.3       gtable_0.3.0     
[33] glue_1.4.2        Rcpp_1.0.5        carData_3.0-4     cellranger_1.1.0 
[37] jquerylib_0.1.4   vctrs_0.3.8       nlme_3.1-149      xfun_0.29        
[41] ps_1.6.0          testthat_3.0.2    rvest_1.0.1       lifecycle_1.0.0  
[45] rstatix_0.7.0     MASS_7.3-53       scales_1.1.1      hms_1.1.0        
[49] promises_1.2.0.1  parallel_4.0.3    yaml_2.2.1        curl_4.3.2       
[53] memoise_2.0.0     sass_0.4.0        stringi_1.5.3     highr_0.9        
[57] desc_1.3.0        pkgbuild_1.2.0    zip_2.1.1         rlang_0.4.10     
[61] pkgconfig_2.0.3   evaluate_0.14     tidyselect_1.1.1  processx_3.5.1   
[65] plyr_1.8.6        R6_2.5.0          generics_0.1.0    DBI_1.1.1        
[69] pillar_1.6.0      haven_2.4.1       whisker_0.4       foreign_0.8-80   
[73] withr_2.4.2       mgcv_1.8-33       abind_1.4-5       modelr_0.1.8     
[77] crayon_1.4.1      car_3.0-11        utf8_1.1.4        grid_4.0.3       
[81] data.table_1.14.0 callr_3.7.0       git2r_0.28.0      reprex_2.0.0     
[85] digest_0.6.27     httpuv_1.6.0      munsell_0.5.0     bslib_0.2.5.1    
[89] sessioninfo_1.1.1</code></pre>
</div>
</div>
</div>

<hr>
<p>
    This site was created with <a href="http://rmarkdown.rstudio.com">R Markdown</a>
</p>
<hr>

<!-- To enable disqus, uncomment the section below and provide your disqus_shortname -->

<!-- disqus
  <div id="disqus_thread"></div>
    <script type="text/javascript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'rmarkdown'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
    <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
    <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
-->

<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
